File: web/src/entities/boss.ts
import { Direction } from '../constants.js';
import { ImmutableVec2 } from '../math.js';
import { lerp } from '../utils.js';
import { Entity } from '../entity.js';
import { PlayerSkill } from '../playerSkill.js';
import { Resource } from '../resource.js';
import { Sound } from '../media/sound.js';
import { Music } from '../media/music.js';
import type { Animation } from '../media/animation.js';
import type { Room } from '../room.js';
import { ParticleSystem } from './particleSystem.js';
import { ReactorCore } from './reactorCore.js';
import { Coin } from './coin.js';

const BLOW_TIME = 60 * 4 + 30;
const DAMAGE_MAX = 16;
const FRAME_PER_DAMAGE = 4;
const INITIAL_BLOW_TIME = 60 * 4 + 30;
const MAX_HEALTH = 40;
const MIN_HEALTH = 10;

const enum BossState {
  INIT, SLEEPING, INITIAL_BLOW, AWAKENING, MOVE_UPWARDS, FLOAT, STOP, ATTACK, VULNERABLE, DEAD
}

export class Boss extends Entity {
  private health = lerp(MAX_HEALTH, MIN_HEALTH, PlayerSkill.get());
  private state: BossState = BossState.INIT;
  private direction: Direction = Direction.LEFT;
  private initialHealth = 16;
  private animFrameCounter = 0;
  private readonly reactorAnim: Animation;
  private readonly bossAnim: Animation;
  private frameCounter = 0;
  private originalPos = new ImmutableVec2();

  constructor() {
    super();
    this.setSize(new ImmutableVec2(30, 40));
    this.reactorAnim = Resource.getAnimation('data/images/reactor_shell.bmp', 4);
    this.bossAnim = Resource.getAnimation('data/images/boss.bmp', 4);
  }

  getLayer(): number { return 3; }
  override isDamagable(): boolean { return true; }

  override getCollideTop(): number { return this.getPosition().y - this.getHalfSize().y - 2; }
  override getCollideLeft(): number { return this.getPosition().x - this.getHalfSize().x - 2; }
  override getCollideBottom(): number { return this.getPosition().y + this.getHalfSize().y + 2; }
  override getCollideRight(): number { return this.getPosition().x + this.getHalfSize().x + 2; }

  override setRoom(room: Room): void {
    super.setRoom(room);
    this.setTilesCollidable(true);
  }

  private setTilesCollidable(v: boolean): void {
    const sx = Math.floor((this.getPosition().x - this.getHalfSize().x) / 10);
    const sy = Math.floor((this.getPosition().y - this.getHalfSize().y) / 10);
    for (let x = sx; x < sx + 3; x++)
      for (let y = sy; y < sy + 4; y++)
        this.myRoom.setCollidable(x, y, v);
  }

  override onDamage(): void {
    if (this.state === BossState.SLEEPING) {
      this.spawnDebris();
      this.initialHealth--;
      if (this.initialHealth === 0) {
        this.frameCounter = 0;
        this.state = BossState.INITIAL_BLOW;
        Sound.playSample('data/sounds/damage');
        this.myRoom.getCamera().addShake(4.0, INITIAL_BLOW_TIME);
      } else {
        Sound.playSample('data/sounds/damage');
        this.myRoom.getCamera().addShake(1.0, 20);
      }
    }

    if (this.state === BossState.VULNERABLE) {
      this.spawnDebris();
      this.health--;
      if (this.health < 0) {
        this.state = BossState.DEAD;
        this.frameCounter = 0;
        Sound.playSample('data/sounds/damage');
        this.myRoom.getCamera().addShake(4.0, BLOW_TIME);
      } else {
        Sound.playSample('data/sounds/damage');
        this.myRoom.getCamera().addShake(1.0, 20);
      }
    }
  }

  private spawnDebris(): void {
    const numPs = Math.abs(Math.floor(Math.random() * 5)) % 5 + 1;
    const ps = new ParticleSystem(
      Resource.getAnimation('data/images/debris.bmp', 4),
      10, 40, 10, 1, 50, numPs, new ImmutableVec2(0, -20), 2.0);
    ps.setPositionWithSpread(this.getPosition(), 10, false);
    this.myRoom.addEntity(ps);
  }

  override onRespawn(): void {
    this.setPosition(this.originalPos);
    this.state = BossState.SLEEPING;
    this.direction = Direction.LEFT;
    this.health = lerp(MAX_HEALTH, MIN_HEALTH, PlayerSkill.get());
    this.initialHealth = 16;
    this.animFrameCounter = 0;
    Music.popSong();
    this.setTilesCollidable(true);
  }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    const x = Math.round(this.getDrawPositionX() + offsetX - this.getHalfSize().x);
    const y = Math.round(this.getDrawPositionY() + offsetY - this.getHalfSize().y);

    if (this.state === BossState.SLEEPING)
      this.reactorAnim.drawFrame(ctx,
        Math.floor(DAMAGE_MAX / FRAME_PER_DAMAGE) - Math.floor((this.initialHealth + 3) / FRAME_PER_DAMAGE), x, y);
    else if (this.state === BossState.INITIAL_BLOW)
      this.reactorAnim.drawFrame(ctx, 3, x, y);
    else if (this.state === BossState.VULNERABLE)
      this.bossAnim.drawFrame(ctx, 0, x, y);
    else
      this.bossAnim.drawFrame(ctx, Math.floor(this.animFrameCounter / 5), x, y);
  }

  override update(): void {
    this.animFrameCounter++;
    this.frameCounter++;

    if (this.state === BossState.INIT) {
      this.originalPos = new ImmutableVec2(this.getPosition());
      this.state = BossState.SLEEPING;
    }

    if (this.state !== BossState.MOVE_UPWARDS && this.state !== BossState.SLEEPING &&
        this.state !== BossState.INITIAL_BLOW && this.state !== BossState.DEAD &&
        this.state !== BossState.VULNERABLE) {
      if (this.myRoom.getHero().Collides(this)) this.myRoom.getHero().kill();
    }

    if (this.state === BossState.INITIAL_BLOW) {
      if (this.frameCounter % 60 === 0) Sound.playSample('data/sounds/alarm');
      if (this.frameCounter >= INITIAL_BLOW_TIME) {
        Sound.playSample('data/sounds/reactor_explosion');
        Sound.playSample('data/sounds/start');
        const ps = new ParticleSystem(
          Resource.getAnimation('data/images/debris.bmp', 4),
          20, 200, 20, 1, 50, 50, new ImmutableVec2(0, -150), 5.0);
        ps.setPositionWithSpread(this.getPosition(), 10, false);
        this.myRoom.addEntity(ps);
        this.frameCounter = 0;
        this.state = BossState.AWAKENING;
      }
    }

    if (this.state === BossState.AWAKENING) {
      if (this.frameCounter > 60 * 3) {
        this.state = BossState.MOVE_UPWARDS;
        this.frameCounter = 0;
        this.setTilesCollidable(false);
        Music.pushSong();
        Music.playSong('data/music/olof3.xm');
      }
    }

    if (this.state === BossState.MOVE_UPWARDS) {
      this.setPosition(new ImmutableVec2(this.getPosition().x, this.getPosition().y - 1));
      if (this.originalPos.y - this.getPosition().y > 130) {
        this.state = BossState.FLOAT;
        this.myRoom.broadcastBoosWallActivate();
        this.frameCounter = 0;
      }
    }

    if (this.state === BossState.FLOAT) {
      const yDelta = Math.sin(this.frameCounter / 8) * 2;
      if (this.direction === Direction.RIGHT) {
        this.setPosition(new ImmutableVec2(this.getPosition().x + 1.5, this.getPosition().y + yDelta));
        if (this.getPosition().x - this.originalPos.x > 110) this.direction = Direction.LEFT;
      }
      if (this.direction === Direction.LEFT) {
        this.setPosition(new ImmutableVec2(this.getPosition().x - 1.5, this.getPosition().y + yDelta));
        if (this.originalPos.x - this.getPosition().x > 110) this.direction = Direction.RIGHT;
      }
      if (this.frameCounter > 60 * 7) {
        this.state = BossState.ATTACK;
        this.frameCounter = 0;
      }
    }

    if (this.state === BossState.ATTACK) {
      if (this.frameCounter < 40) {
        if (this.frameCounter % 2 === 0)
          this.setPosition(new ImmutableVec2(this.getPosition().x - 2, this.getPosition().y));
        else
          this.setPosition(new ImmutableVec2(this.getPosition().x + 2, this.getPosition().y));
      } else {
        this.setPosition(new ImmutableVec2(this.getPosition().x, this.getPosition().y + 6));
        if (this.originalPos.y < this.getPosition().y) {
          this.myRoom.broadcastBoosWallDectivate();
          this.setPosition(new ImmutableVec2(this.getPosition().x, this.originalPos.y));
          this.state = BossState.VULNERABLE;
          this.myRoom.getCamera().addShake(5.0, 80);

          const numPs = Math.abs(Math.floor(Math.random() * 10)) % 10 + 5;
          const ps = new ParticleSystem(
            Resource.getAnimation('data/images/debris.bmp', 4),
            10, 40, 10, 1, 50, numPs, new ImmutableVec2(0, -20), 2.0);
          ps.setPositionWithSpread(this.getPosition(), 10, false);
          this.myRoom.addEntity(ps);
          this.myRoom.broadcastBoosFloorActivate();
          Sound.playSample('data/sounds/reactor_explosion');
          Coin.spawnDeathCoins(
            Math.abs(Math.floor(Math.random() * 5)) % 5 + 3,
            this.getPosition(), 60 * 6, this.myRoom);
        }
      }
    }

    if (this.state === BossState.VULNERABLE) {
      if (this.frameCounter > 60 * 5) {
        this.state = BossState.MOVE_UPWARDS;
        this.frameCounter = 0;
      }
    }

    if (this.state === BossState.DEAD) {
      if (this.frameCounter % 60 === 0) Sound.playSample('data/sounds/alarm');
      if (this.frameCounter >= BLOW_TIME) {
        Sound.playSample('data/sounds/reactor_explosion');
        Sound.playSample('data/sounds/start');
        const ps = new ParticleSystem(
          Resource.getAnimation('data/images/debris.bmp', 4),
          20, 200, 20, 1, 50, 50, new ImmutableVec2(0, -150), 5.0);
        ps.setPositionWithSpread(this.getPosition(), 10, false);
        this.myRoom.addEntity(ps);
        const core = new ReactorCore();
        core.setPosition(this.getPosition());
        core.setVelocity(new ImmutableVec2(0, -50));
        this.myRoom.addEntity(core);
        this.remove();
      }
    }
  }
}

File: wii/src/entities/boss.cpp
#include "entities/boss.h"
#include "entities/particle_system.h"
#include "entities/reactor_core.h"
#include "entities/coin.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "media/sound.h"
#include "media/music.h"
#include "room.h"
#include "camera.h"
#include "player_skill.h"
#include "utils.h"
#include <cmath>
#include <cstdlib>

static const int BLOW_TIME = 60 * 4 + 30;
static const int DAMAGE_MAX = 16;
static const int FRAME_PER_DAMAGE = 4;
static const int INITIAL_BLOW_TIME = 60 * 4 + 30;
static const float MAX_HEALTH = 40.0f;
static const float MIN_HEALTH = 10.0f;

Boss::Boss()
    : health_(lerp(MAX_HEALTH, MIN_HEALTH, PlayerSkill::get())) {
    setSize(Vec2(30, 40));
    reactorAnim_ = Resource::getAnimation("data/images/reactor_shell.bmp", 4);
    bossAnim_ = Resource::getAnimation("data/images/boss.bmp", 4);
}

void Boss::setRoom(Room* room) {
    Entity::setRoom(room);
    setTilesCollidable(true);
}

void Boss::setTilesCollidable(bool v) {
    int sx = static_cast<int>(std::floor((position_.x - halfSize_.x) / 10.0f));
    int sy = static_cast<int>(std::floor((position_.y - halfSize_.y) / 10.0f));
    for (int x = sx; x < sx + 3; x++)
        for (int y = sy; y < sy + 4; y++)
            room_->setCollidable(x, y, v);
}

void Boss::spawnDebris() {
    int numPs = std::abs(rand() % 5) + 1;
    ParticleSystem* ps = new ParticleSystem(
        Resource::getAnimation("data/images/debris.bmp", 4),
        10, 40, 10, 1.0f, 50.0f, numPs, Vec2(0, -20), 2.0f);
    ps->setPositionWithSpread(position_, 10.0f, false);
    room_->addEntity(ps);
}

void Boss::onDamage() {
    if (state_ == BossState::SLEEPING) {
        spawnDebris();
        initialHealth_--;
        if (initialHealth_ == 0) {
            bossFrameCounter_ = 0;
            state_ = BossState::INITIAL_BLOW;
            Sound::playSample("data/sounds/damage");
            room_->getCamera()->addShake(4.0f, static_cast<float>(INITIAL_BLOW_TIME));
        } else {
            Sound::playSample("data/sounds/damage");
            room_->getCamera()->addShake(1.0f, 20.0f);
        }
    }

    if (state_ == BossState::VULNERABLE) {
        spawnDebris();
        health_ -= 1.0f;
        if (health_ < 0) {
            state_ = BossState::DEAD;
            bossFrameCounter_ = 0;
            Sound::playSample("data/sounds/damage");
            room_->getCamera()->addShake(4.0f, static_cast<float>(BLOW_TIME));
        } else {
            Sound::playSample("data/sounds/damage");
            room_->getCamera()->addShake(1.0f, 20.0f);
        }
    }
}

void Boss::onRespawn() {
    setPosition(originalPos_);
    state_ = BossState::SLEEPING;
    direction_ = Direction::LEFT;
    health_ = lerp(MAX_HEALTH, MIN_HEALTH, PlayerSkill::get());
    initialHealth_ = 16;
    animFrameCounter_ = 0;
    Music::popSong();
    setTilesCollidable(true);
}

void Boss::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(getDrawPositionX() + offsetX - halfSize_.x));
    int y = static_cast<int>(std::round(getDrawPositionY() + offsetY - halfSize_.y));

    if (state_ == BossState::SLEEPING) {
        int frame = DAMAGE_MAX / FRAME_PER_DAMAGE - (initialHealth_ + 3) / FRAME_PER_DAMAGE;
        reactorAnim_->drawFrame(renderer, frame, x, y);
    } else if (state_ == BossState::INITIAL_BLOW) {
        reactorAnim_->drawFrame(renderer, 3, x, y);
    } else if (state_ == BossState::VULNERABLE) {
        bossAnim_->drawFrame(renderer, 0, x, y);
    } else {
        bossAnim_->drawFrame(renderer, animFrameCounter_ / 5, x, y);
    }
}

void Boss::update() {
    animFrameCounter_++;
    bossFrameCounter_++;

    if (state_ == BossState::INIT) {
        originalPos_ = position_;
        state_ = BossState::SLEEPING;
    }

    if (state_ != BossState::MOVE_UPWARDS && state_ != BossState::SLEEPING &&
        state_ != BossState::INITIAL_BLOW && state_ != BossState::DEAD &&
        state_ != BossState::VULNERABLE) {
        if (room_->getHero()->Collides(*this)) room_->getHero()->kill();
    }

    if (state_ == BossState::INITIAL_BLOW) {
        if (bossFrameCounter_ % 60 == 0) Sound::playSample("data/sounds/alarm");
        if (bossFrameCounter_ >= INITIAL_BLOW_TIME) {
            Sound::playSample("data/sounds/reactor_explosion");
            Sound::playSample("data/sounds/start");
            ParticleSystem* ps = new ParticleSystem(
                Resource::getAnimation("data/images/debris.bmp", 4),
                20, 200, 20, 1.0f, 50.0f, 50, Vec2(0, -150), 5.0f);
            ps->setPositionWithSpread(position_, 10.0f, false);
            room_->addEntity(ps);
            bossFrameCounter_ = 0;
            state_ = BossState::AWAKENING;
        }
    }

    if (state_ == BossState::AWAKENING) {
        if (bossFrameCounter_ > 60 * 3) {
            state_ = BossState::MOVE_UPWARDS;
            bossFrameCounter_ = 0;
            setTilesCollidable(false);
            Music::pushSong();
            Music::playSong("data/music/olof3.xm");
        }
    }

    if (state_ == BossState::MOVE_UPWARDS) {
        setPosition(Vec2(position_.x, position_.y - 1.0f));
        if (originalPos_.y - position_.y > 130.0f) {
            state_ = BossState::FLOAT;
            room_->broadcastBossWallActivate();
            bossFrameCounter_ = 0;
        }
    }

    if (state_ == BossState::FLOAT) {
        float yDelta = std::sin(bossFrameCounter_ / 8.0f) * 2.0f;
        if (direction_ == Direction::RIGHT) {
            setPosition(Vec2(position_.x + 1.5f, position_.y + yDelta));
            if (position_.x - originalPos_.x > 110.0f) direction_ = Direction::LEFT;
        }
        if (direction_ == Direction::LEFT) {
            setPosition(Vec2(position_.x - 1.5f, position_.y + yDelta));
            if (originalPos_.x - position_.x > 110.0f) direction_ = Direction::RIGHT;
        }
        if (bossFrameCounter_ > 60 * 7) {
            state_ = BossState::ATTACK;
            bossFrameCounter_ = 0;
        }
    }

    if (state_ == BossState::ATTACK) {
        if (bossFrameCounter_ < 40) {
            if (bossFrameCounter_ % 2 == 0)
                setPosition(Vec2(position_.x - 2.0f, position_.y));
            else
                setPosition(Vec2(position_.x + 2.0f, position_.y));
        } else {
            setPosition(Vec2(position_.x, position_.y + 6.0f));
            if (originalPos_.y < position_.y) {
                room_->broadcastBossWallDeactivate();
                setPosition(Vec2(position_.x, originalPos_.y));
                state_ = BossState::VULNERABLE;
                room_->getCamera()->addShake(5.0f, 80.0f);

                int numPs = std::abs(rand() % 10) + 5;
                ParticleSystem* ps = new ParticleSystem(
                    Resource::getAnimation("data/images/debris.bmp", 4),
                    10, 40, 10, 1.0f, 50.0f, numPs, Vec2(0, -20), 2.0f);
                ps->setPositionWithSpread(position_, 10.0f, false);
                room_->addEntity(ps);
                room_->broadcastBossFloorActivate();
                Sound::playSample("data/sounds/reactor_explosion");
                int coinCount = std::abs(rand() % 5) + 3;
                Coin::spawnDeathCoins(coinCount, position_, 60 * 6, room_);
            }
        }
    }

    if (state_ == BossState::VULNERABLE) {
        if (bossFrameCounter_ > 60 * 5) {
            state_ = BossState::MOVE_UPWARDS;
            bossFrameCounter_ = 0;
        }
    }

    if (state_ == BossState::DEAD) {
        if (bossFrameCounter_ % 60 == 0) Sound::playSample("data/sounds/alarm");
        if (bossFrameCounter_ >= BLOW_TIME) {
            Sound::playSample("data/sounds/reactor_explosion");
            Sound::playSample("data/sounds/start");
            ParticleSystem* ps = new ParticleSystem(
                Resource::getAnimation("data/images/debris.bmp", 4),
                20, 200, 20, 1.0f, 50.0f, 50, Vec2(0, -150), 5.0f);
            ps->setPositionWithSpread(position_, 10.0f, false);
            room_->addEntity(ps);
            ReactorCore* core = new ReactorCore();
            core->setPosition(position_);
            core->setVelocity(Vec2(0, -50));
            room_->addEntity(core);
            remove();
        }
    }
}
